src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt

```diff
@@ -43,15 +70,30 @@ interface ICriterionTableRepo {
 class CriterionTableRepo(
     private val db: IDatabase,
 ) : ICriterionTableRepo {
+    /**
+     * Requesting a criterion from the database.
+     *
+     * @param id The id of the requested criterion.
+     * @return The [Criterion] object or null, if no criterion with the given [id] was found.
+     */
+    private fun getCriterionByIdOrNull(id: UUID): Criterion? = CriterionTable
+        .selectAll()
+        .where { CriterionTable.id eq id }
+        .map { it.toCriterion() }
+        .singleOrNull()
+
+    override suspend fun getCriterionById(id: UUID): Criterion = db.dbQuery {
+        getCriterionByIdOrNull(id) ?: throw NotFoundException(EntityType.CRITERION, id.toString())
+    }
+
     override suspend fun createCriterion(request: CriterionOuterClass.Criterion.Create, userId: UUID): Criterion =
         db.dbQuery {
-            val projectUUID = parseUUID(request.projectId, EntityType.PROJECT)
-
-            // Get user reference
             val userEntityId = getUserEntityId(userId)
-
-            // Get project reference
-            val projectEntityId = getProjectEntityId(projectUUID)
+            var projectEntityId: EntityID<UUID>? = null
+            if (!request.projectId.isNullOrEmpty()) {
+                val projectUUID = parseUUID(request.projectId, EntityType.PROJECT)
+                projectEntityId = getProjectEntityId(projectUUID)
+            }
```

Comment:
The `getUserEntityId` and `getProjectEntityId` calls are obsolete. I initially thought we would require the `EntityID` object, but we can simply use the wrapped value. Additionally, we only need to check for nullability because we assume that the projectId is not empty.

In the end we can reduce this to a single line.

```suggestion
        db.dbQuery {
            val projectId = request.projectId?.let { parseUUID(it, EntityType.PROJECT) }
```

Further down, we just have to adjust the used IDs:
```kotlin
it[CriterionTable.projectId] = projectId
it[createdBy] = userId
```

Reply:
I don't think this is correct. I tried it a couple times now and I get an referential integrity error if remove the `getProjectEntityId` or the `getUserEntitiyId`. I did not find any solution without those that worked.

Reply:
The second point is with the not null check. I don't think this is right either. As far as I know if the `project_id` field of an grpc call is not defined it is set empty. This behaviour can also be seen in the tests, if we don't set the `project_id`. Therefore we can only remove the nullability check but not the emptiness check.

Reply:
Interesting, I thought we had not used the entityId references in other repo methods, but apparently we didn't.